src/engraving/rendering/dev/accidentalslayout.cpp

```diff
@@ -0,0 +1,1118 @@
+        for (auto noteSplitGroup : noteSplitGroups) {
+            std::vector<std::vector<Accidental*> > accidSplitGroups = splitAccordingToAccidDisplacement(noteSplitGroup, ctx);
+            groups.insert(groups.end(), accidSplitGroups.begin(), accidSplitGroups.end());
+        }
+    } else {
+        groups = splitAccordingToAccidDisplacement(accidentals, ctx);
+    }
+
+    size_t curGroupsCount = groups.size();
+    if (ctx.alignOctavesAcrossSubChords()) {
+        mergeSubGroupsWithOctavesAcross(ctx);
+        if (groups.size() != curGroupsCount) {
+            for (auto& subGroup : groups) {
+                sortTopDown(subGroup);
+            }
+        }
+    }
+
+    return groups;
+}
+
+std::vector<std::vector<Accidental*> > AccidentalsLayout::splitAccordingToAccidDisplacement
+    (std::vector<Accidental*>& accidentals, AccidentalsLayoutContext& ctx)
+{
+    std::vector<std::vector<Accidental*> > subGroups = groupAccidentalsByXPos(accidentals, ctx);
+
+    muse::remove_if(subGroups, [](auto& subGroup) {
+        return subGroup.size() == 0;
+    });
+
+    if (subGroups.size() > 1 && subGroups[0].size() > 1) {
+        moveOctavesToSecondGroup(subGroups);
+    }
+
+    for (std::vector<Accidental*>& subGroup : subGroups) {
+        sortTopDown(subGroup);
+    }
+
+    return subGroups;
+}
+
+std::vector<std::vector<Accidental*> > AccidentalsLayout::groupAccidentalsByXPos
+    (std::vector<Accidental*>& accidentals, AccidentalsLayoutContext& ctx)
+{
+    std::map<double, std::vector<Accidental*> > groupsOfEqualX;
+
+    for (Accidental* acc : accidentals) {
+        Shape accRoughShape = Shape(acc->symBbox(acc->symId()), acc);
+        accRoughShape.translateY(acc->note()->y());
+        double x = -minAccidentalToChordDistance(acc, accRoughShape, ctx) + acc->symWidth(acc->symId());
+        groupsOfEqualX[std::round(x)].push_back(acc);
+    }
+
+    double splitPoint = DBL_MAX;
+    double prevX = (*groupsOfEqualX.begin()).first;
+    for (auto& pair : groupsOfEqualX) {
+        double curX = pair.first;
+        if (curX - prevX > 0.5 * ctx.spatium()) {
+            splitPoint = curX;
+            break;
+        }
+        prevX = curX;
+    }
+
+    std::vector<std::vector<Accidental*> > subGroups(2);
+    for (auto& pair : groupsOfEqualX) {
+        if (pair.first >= splitPoint) {
+            for (Accidental* acc : pair.second) {
+                subGroups[0].push_back(acc);
+                acc->mutldata()->verticalSubgroup = 0;
+            }
+        } else {
+            for (Accidental* acc : pair.second) {
+                subGroups[1].push_back(acc);
+                acc->mutldata()->verticalSubgroup = 1;
+            }
+        }
+    }
+
+    return subGroups;
+}
+
+void AccidentalsLayout::moveOctavesToSecondGroup(std::vector<std::vector<Accidental*> >& subGroups)
+{
+    std::vector<Accidental*>& firstGroup = subGroups[0];
+    std::vector<Accidental*>& secondGroup = subGroups[1];
```

Comment:
It looks like std::array might be a better type for subGroups

Reply:
This is one particular case (with conditions defined by the caller), but in general subGroups may have different sizes

Reply:
worth adding this

IF_ASSERT_FAILED(subGroups.size() == 2) {
    return;
}